crates/task-impls/src/quorum_proposal_recv.rs

```diff
@@ -0,0 +1,206 @@
+
+    /// Network for all nodes
+    pub quorum_network: Arc<I::QuorumNetwork>,
+
+    /// Membership for Quorum Certs/votes
+    pub quorum_membership: Arc<TYPES::Membership>,
+
+    /// Membership for Timeout votes/certs
+    pub timeout_membership: Arc<TYPES::Membership>,
+
+    /// timeout task handle
+    pub timeout_task: Option<JoinHandle<()>>,
+
+    /// View timeout from config.
+    pub timeout: u64,
+
+    /// Round start delay from config, in milliseconds.
+    pub round_start_delay: u64,
+
+    /// Output events to application
+    pub output_event_stream: async_broadcast::Sender<Event<TYPES>>,
+
+    /// This node's storage ref
+    pub storage: Arc<RwLock<I::Storage>>,
+
+    /// The most recent upgrade certificate this node formed.
+    /// Note: this is ONLY for certificates that have been formed internally,
+    /// so that we can propose with them.
+    ///
+    /// Certificates received from other nodes will get reattached regardless of this fields,
+    /// since they will be present in the leaf we propose off of.
+    pub formed_upgrade_certificate: Option<UpgradeCertificate<TYPES>>,
+
+    /// last View Sync Certificate or Timeout Certificate this node formed.
+    pub proposal_cert: Option<ViewChangeEvidence<TYPES>>,
+
+    /// most recent decided upgrade certificate
+    pub decided_upgrade_cert: Option<UpgradeCertificate<TYPES>>,
+
+    /// Spawned tasks related to a specific view, so we can cancel them when
+    /// they are stale
+    pub spawned_tasks: BTreeMap<TYPES::Time, Vec<JoinHandle<()>>>,
+
+    /// Immutable instance state
+    pub instance_state: Arc<TYPES::InstanceState>,
+
+    /// The node's id
+    pub id: u64,
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> QuorumProposalRecvTaskState<TYPES, I> {
+    /// Cancel all tasks the consensus tasks has spawned before the given view
+    pub async fn cancel_tasks(&mut self, view: TYPES::Time) {
+        let keep = self.spawned_tasks.split_off(&view);
+        let mut cancel = Vec::new();
+        while let Some((_, tasks)) = self.spawned_tasks.pop_first() {
+            let mut to_cancel = tasks.into_iter().map(cancel_task).collect();
+            cancel.append(&mut to_cancel);
+        }
+        self.spawned_tasks = keep;
+        join_all(cancel).await;
+    }
+
+    /// Handles all consensus events relating to propose and vote-enabling events.
+    #[instrument(skip_all, fields(id = self.id, view = *self.cur_view), name = "Consensus replica task", level = "error")]
+    #[allow(unused_variables)]
+    pub async fn handle(
+        &mut self,
+        event: Arc<HotShotEvent<TYPES>>,
+        event_stream: Sender<Arc<HotShotEvent<TYPES>>>,
+    ) {
+        #[cfg(feature = "dependency-tasks")]
+        if let HotShotEvent::QuorumProposalRecv(proposal, sender) = event.as_ref() {
+            match handle_quorum_proposal_recv(proposal, sender, event_stream.clone(), self).await {
+                Ok(Some(current_proposal)) => {
```

Comment:
As mentioned offline, I think we should call `cancel_tasks` for this view if the result is `Ok`.

Otherwise, looks like we never cancel the spawned tasks and can just remove the `cancel_tasks` function.

Reply:
Done